### ISSUE TITLE
RFC: Design of Configuration File

### ISSUE THREAD

**Issue by Timidger** [way-cooler/way-cooler#219]
Continuing the conversation from #215, this is a request for others to contribute their opinion to the theoretical design for the configuration file in Way Cooler.

tldr: Most configuration options will be in a user-controlled ~yaml~ toml file, but client permissions and keybindings will be in a root controlled `init.lua` file.

> I've thought about this a lot, and here's my reasoning (and feel free to tell my this is a mistake -- it's possible it is!)
> 
> Both Wayland and Rust are designed around the idea that security is very important. Too much software these days doesn't take this into account. This is especially important for something with a large attack vector as a window manager designed to be as extendable by the user as possible.
> 
> The main way we will control the security of the window manager is by assigning permissions to client programs through the `init.lua`. Clients will need to request permission for things that in X11 you could just do freely (such as kill windows, get information about other windows, get pointer position, etc.). Sway does something similar, and we would do the same thing but via D-Bus and Lua.
> 
> However, if the `init.lua` file is in a path that is writeable by the user, a malicious script could inject permissions into the file and basically gain control of the window manager. 
> 
> My solution is to move `init.lua` into `/etc/` and move *most* of the configuration options into a ~yaml~ toml file in `$XDG_HOME/way-cooler/config.toml`. E.g colors, border size, gap size, etc. Basically the things that users would want to change on a regular basis.
> 
> However, both keybindings and client permissions would live in `/etc/init.lua`. They are the largest attack vectors, and also the things that are most rarely changed. Keybindings _needs_ to be in a Lua file because you can bind lua functions to keybindings.
> 
> My reasoning is if the user really wants a Lua file in a user writable path, they can add `require("/path/to/insecure/lua/file.lua")` at the end of their `/etc/way-cooler/init.lua` file.
> 
> I think it's possible to remain highly configurable, but also be very secure at the same time with this setup.
> 
> I have come up with a few alternative, but they are more complicated and also less secure:
> 
> * Having two lua files, one for client permissions (in `/etc/`) and another for keybindings (in `$XDG_HOME`)
> * Caching permissions in a root file somewhere and prompting the user when a new one is added.

**Comment by coder543** [comment]
Starting with one of the fundamental assumptions: Do you really need a Turing-complete language for a configuration file? It seems like everything could be described in `yaml` (or maybe even better, `toml`), even binding key presses to named `lua` functions, or specifying `lua` files that you want to be executed at initialization.

**Comment by coder543** [comment]
I also think a sane option would be that if no `/etc/init.lua` file exists, Way Cooler should search for such a file in a specific location within the user's home directory. If the user wants to be security conscious, they can put the `lua` file in `/etc/`, and no malicious program will be able to mess with it, but Way Cooler will fallback to a home directory `init.lua` if no such `/etc/` file exists.

**Comment by Timidger** [comment]
> Starting with one of the fundamental assumptions: Do you really need a Turing-complete language for a configuration file? 

We wanted users to be able to do arbitrary things with their window manager / applications, whether at startup time or at some other time, without having to do a weird hack or make a client program. This is similar in design to awesome.

>It seems like everything could be described in yaml (or maybe even better, toml)

Gah, I meant to write toml. I don't know why I wrote yaml. I'll fix that in the op

> I also think a sane option would be that if no /etc/init.lua file exists, Way Cooler should search for such a file in a specific location within the user's home directory. If the user wants to be security conscious, they can put the lua file in /etc/, and no malicious program will be able to mess with it, but Way Cooler will fallback to a home directory init.lua if no such /etc/ file exists, since some users want convenience.

Hmm, that's interesting because that's basically the reverse of both how Way Cooler currently works and what is expected of most programs. Generally the configuration files in `/etc/` are system defined, with user defined ones over-riding the system defined ones. It's a possibility I hadn't considered though, and perhaps a better compromise than the ones I laid out.

**Comment by Stebalien** [comment]
I think you're misunderstanding wayland security. Wayland security *does not* protect the users from unrestricted applications. Instead, wayland allows users to safely sandbox applications (and/or run them as another user). Currently any application running as any user that can talk to some Xorg instance can use Xorg to effectively escalate privileges (by messing with other Xorg windows). 

Under wayland, applications cannot mess with each other *through* wayland. However, unrestricted applications can certainly mess with the user's files and, at that point, messing with way-cooler's config is the least of your concerns. They can replace the bash profile, mess with environment variables (\*cough\* `LD_PRELOAD` \*cough\*), and, really, do whatever they want.

**Comment by Timidger** [comment]
@stebalien 

Those are good points, at a certain point security does have to reach a point where it is "good enough" with regards to the components and system that it is a part of. Way Cooler certainly can't expect every possibility of malicious attack (like LD_PRELOAD hacks ;-)). 

I understand Wayland wasn't designed as a catch all to stop bad programs forever. By "default" (eg with the most basic of compositors) each application is very sand boxed from other applications, and it's up to the compositor to allow them to communicate / modify things outside of the container's little world. 

Obviously this sand boxing only applies to the window manager and there are still scores of ways to mess with the user. I was hoping however that the user could use Way Cooler with the assurance that, at the very least, Way Cooler itself can not be compromised (ignoring things like LD_PRELOAD hacking shared libraries). 

But, ultimately,  I think you are correct -- I'm being probably being over zealous here. The user can already secure their files (either with root or some complicated multi user setup) and Way Cooler would be secure enough with just client permissions.

**Comment by thefarwind** [comment]
I'd prefer the key-bindings to remain per user, rather than being located in an /etc/ file. I agree with the sentiment that if an attacker were able to get in and modify the init file, they're already going to be able to do far worse.

**Comment by Timidger** [comment]
@thefarwind also a good point, that would break the use case of multi-user setups.

**Comment by Timidger** [comment]
Thank you all for your feedback!

I've discussed this with @SnirkImmington, and we have decided to not require that `init.lua` be in `/etc/`. We will continue to read the configuration file in `$XDG_HOME/way-cooler/`, defaulting back to `/etc/` if not found (and defaulting to the compiled-in init file if that cannot be found as well).

With that decision, all configuration is going to remain in the Lua file. We will add support for loading "theme" files described as yaml (not toml, I apologize for the confusion I caused) from the Lua file, but all of the configuration will continue to use and execute Lua at startup. We haven't thought about how these theme files will be structured, but we are narrowing the choices down between yaml and toml. More details will appear later, once we actually have something to theme ;)
